src/imitation/data/types.py

```diff
@@ -22,7 +29,267 @@
+
+        Raises:
+            RuntimeError: if the arrays have different lengths or there are no arrays.
+
+        Returns:
+            The length (first dimension) of the constiuent arrays
+        """
+        lens = set(len(v) for v in self._d.values())
+        if len(lens) == 1:
+            return lens.pop()
+        elif len(lens) == 0:
+            raise RuntimeError("Length not defined as DictObs is empty")
+        else:
+            raise RuntimeError(
+                f"Length not defined; arrays have conflicting first dimensions: {lens}",
+            )
+
+    @property
+    def dict_len(self):
+        """Returns the number of arrays in the DictObs."""
+        return len(self._d)
+
+    def __getitem__(
+        self,
+        key: Union[int, slice, Tuple[Union[int, slice], ...]],
+    ) -> "DictObs":
+        """Indexes or slices each array.
+
+        See `.get` for accessing a value from the underlying dictionary.
+
+        Note that it will still return singleton values as np.arrays, not scalars,
+        to be consistent with DictObs type signature.
+
+        Args:
+            key: a single slice
+
+        Returns:
+            A new DictObj object with each array indexed.
+        """
+        # asarray handles case where we slice to a single array element.
+        return self.__class__({k: np.asarray(v[key]) for k, v in self._d.items()})
+
+    def __iter__(self) -> Iterator["DictObs"]:
+        """Iterates over the first dimension of each array.
+
+        Raises:
+            ValueError if len() is not defined.
+
+        Returns:
+            Iterator of dictobjs by first dimension.
+        """
+        return (self[i] for i in range(len(self)))
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            return False
+        if not self.keys() == other.keys():
+            return False
+        return all(np.array_equal(self.get(k), other.get(k)) for k in self.keys())
+
+    @property
+    def shape(self) -> Dict[str, Tuple[int, ...]]:
+        """Returns a dictionary with shape-tuples in place of the arrays."""
+        return {k: v.shape for k, v in self.items()}
+
+    @property
+    def dtype(self) -> Dict[str, np.dtype]:
+        """Returns a dictionary with dtype-tuples in place of the arrays."""
+        return {k: v.dtype for k, v in self.items()}
+
+    def keys(self):
+        return self._d.keys()
+
+    def values(self):
+        return self._d.values()
+
+    def items(self):
+        return self._d.items()
+
+    def __contains__(self, key):
+        return key in self._d
+
+    def get(self, key: str) -> np.ndarray:
+        """Returns the array for the given key, or raises KeyError."""
+        return self._d[key]
+
+    def unwrap(self) -> Dict[str, np.ndarray]:
+        """Returns a copy of the underlying dictionary (arrays are not copied)."""
+        return {k: v for k, v in self._d.items()}
```

Comment:
I think this is equivalent to:
```suggestion
        return dict(self._d)
```